working/0723-static-extensions/feature-specification-variant2.md

```diff
@@ -0,0 +1,347 @@
+are always ignored by static member declarations, they are only relevant to
+constructors.*
+
+When an extension declaration _D_ has no formal type parameters, and
+it has an on-type `C<S1 .. Sk>`, the instantiated constructor return type
+of _D_ is `C<S1 .. Sk>`. *In this case the on-type is a fixed type (also
+known as a ground type), e.g., `List<int>`. This implies that the
+constructor return type of D is the same for every call site.*
+
+Finally we have the general case: Consider an extension declaration
+_D_ named `E` with formal type parameters `X1 extends B1 .. Xs extends Bs`
+and a constructor return type `C<S1 .. Sk>`. With actual type arguments
+`T1 .. Ts`, the instantiated constructor return type of _D_ with type
+arguments `T1 .. Ts` is `[T1/X1 .. Ts/Xs]C<S1 .. Sk>`.
+
+#### Explicit invocation of a constructor in an extension
+
+Explicit constructor invocations are similar to static member invocations,
+but they need more detailed rules because they can use the formal type
+parameters declared by an extension.
+
+An _explicitly resolved invocation_ of a constructor named `C.name` in a
+extension declaration _D_ named `E` with `s` type parameters and on-class
+`C` can be expressed as `E<S1 .. Ss>.C.name(args)`, 
+`E.C<U1 .. Uk>.name(args)`, or `E<S1 .. Ss>.C<U1 .. Uk>.name(args)` (and
+similarly for a constructor named `C` using `E<S1 .. Ss>.C(args)`, etc).
+
+A compile-time error occurs if the type arguments passed to `E` violate the
+declared bounds. A compile-time error occurs if no type arguments are
+passed to `E`, and type arguments `U1 .. Uk` are passed to `C`, but no list
+of actual type arguments for the type variables of `E` can be found such
+that the instantiated constructor return type of `E` with said type
+arguments is `C<U1 .. Uk>`.
+
+*Note that we must be able to choose the values of the type parameters
+`X1 .. Xs` such that the instantiated constructor return type is
+exactly `C<U1 .. Uk>`, it is not sufficient that it is a subtype thereof,
+or that it differs in any other way.*
+
+A compile-time error occurs if the invocation passes actual type arguments
+to both `E` and `C`, call them `S1 .. Ss` and `U1 .. Uk`, respectively,
+unless the instantiated constructor return type of _D_ with actual type
+arguments `S1 .. Ss` is `C<U1 .. Uk>`. In this type comparison, top types
+like `dynamic` and `Object?` are considered different, and no type
+normalization occurs. *In other words, the types must be equal, not
+just mutual subtypes.*
+
+*Note that explicitly resolved invocations of constructors declared in
+extensions are an exception in real code, usable in the case where a
+name clash prevents an implicitly resolved invocation. However, implicitly
+resolved invocations are specified in the rest of this section by reducing
+them to explicitly resolved ones.*
+
+A constructor invocation of the form `C<T1 .. Tm>.name(args)` is partially
+resolved by looking up a constructor named `C.name` in the class `C` and in
```

Comment:
This looks explicitly for constructors. It probably should look for any static member, and eventually make it an error if there is only one and it's not a constructor.

I know `C<T...>.name` must be a constructor invocation, but (a) I don't want to hardcode that in case we want "class members" at some point, and (b) it differs from `C.name`, even though both can denote a constructor.

Which means that we can't *generally* spearate lookups into member and constructors, we have to be able to handle `C.name(args)`, which can be either,
and I don't want to make it *better* to write `C<int>.name(args)` than `C.name(args)`-which-infers-`<int>`.

No overloading! :grin:

Reply:
I created https://github.com/dart-lang/language/issues/4049 in order to take this discussion separately.